core/src/main/java/org/apache/calcite/runtime/Utilities.java

```diff
@@ -212,6 +213,27 @@ public static int compareNullsLast(Comparable v0, Comparable v1) {
                 : v0.compareTo(v1);
   }
 
+  public static int compare(Comparable v0, Comparable v1, Comparator comparator) {
```

Comment:
Do we need all this new variants? Note that there are also the standard java methods `Comparator.nullsFirst` and `Comparator.nullsLast`. 

If we have a `Comparator` at hand then we could directly generate the code:
```
Comparator nfc = Comparator.nullsFirst(comparator);
nfc.compare(v0, v1);
``` 
I don't see clearly why we need to use the `Utilties` redirection in this case.

Reply:
These methods help us to have a compact and homogeneous code in `PhysTypeImpl.java`, where "complex" comparator expressions can be generated, combining e.g. numeric keys (without comparator), with String keys (with comparator).